synedrion/src/cggmp21/key_init_tests.rs

```diff
@@ -0,0 +1,187 @@
+        entry_points.clone(),
+        2,
+        CheckPart::NormalBroadcast,
+        &(),
+        false,
+    )
+    .unwrap();
+    check_invalid_message_evidence::<SP, _>(
+        &mut OsRng,
+        entry_points.clone(),
+        3,
+        CheckPart::NormalBroadcast,
+        &(),
+        true,
+    )
+    .unwrap();
+
+    check_invalid_message_evidence::<SP, _>(
+        &mut OsRng,
+        entry_points.clone(),
+        1,
+        CheckPart::DirectMessage,
+        &(),
+        false,
+    )
+    .unwrap();
+    check_invalid_message_evidence::<SP, _>(
+        &mut OsRng,
+        entry_points.clone(),
+        2,
+        CheckPart::DirectMessage,
+        &(),
+        false,
+    )
+    .unwrap();
+    check_invalid_message_evidence::<SP, _>(
+        &mut OsRng,
+        entry_points.clone(),
+        3,
+        CheckPart::DirectMessage,
+        &(),
+        false,
+    )
+    .unwrap();
+}
+
+#[test]
+fn r2_hash_mismatch() {
+    struct Override;
+
+    impl Misbehaving<Id, ()> for Override {
+        type EntryPoint = KeyInit<P, Id>;
+
+        fn modify_echo_broadcast(
+            rng: &mut impl CryptoRngCore,
+            round: &BoxedRound<Id, <Self::EntryPoint as EntryPoint<Id>>::Protocol>,
+            _behavior: &(),
+            serializer: &Serializer,
+            deserializer: &Deserializer,
+            echo_broadcast: EchoBroadcast,
+        ) -> Result<EchoBroadcast, LocalError> {
+            if round.id() == 2 {
+                let orig_message = echo_broadcast
+                    .deserialize::<Round2EchoBroadcast<P>>(deserializer)
+                    .unwrap();
+                let mut data = orig_message.data;
+
+                // Replace `u` with something other than we committed to when hashing it in Round 1.
+                data.u = BitVec::random(rng, data.u.bits().len());
+
+                let message = Round2EchoBroadcast { data };
+                return EchoBroadcast::new(serializer, message);
+            }
+
+            Ok(echo_broadcast)
+        }
+    }
+
+    check_evidence::<Override>("Protocol error: A hash mismatch in Round 2").unwrap();
+}
+
+#[test]
+fn r3_invalid_sch_proof() {
+    struct Override;
+
+    impl Misbehaving<Id, ()> for Override {
+        type EntryPoint = KeyInit<P, Id>;
+
+        fn modify_normal_broadcast(
+            rng: &mut impl CryptoRngCore,
+            round: &BoxedRound<Id, <Self::EntryPoint as EntryPoint<Id>>::Protocol>,
+            _behavior: &(),
+            serializer: &Serializer,
+            _deserializer: &Deserializer,
+            normal_broadcast: NormalBroadcast,
+        ) -> Result<NormalBroadcast, LocalError> {
+            if round.id() == 3 {
+                let round3 = round.downcast_ref::<Round3<P, Id>>()?;
+                let context = &round3.context;
+                let aux = (&context.sid_hash, &context.my_id, &round3.rho);
+
+                // Make a proof for a random secret. This won't pass verification.
```

Comment:
These comments are very useful.

Reply:
I'll add more on the second pass, I was just a little tired of writing all these tests :)